loader: explicitly manage policy programs in the Cilium agent

ti-mo opened the pull request:
Fixes: #29331

This PR puts the agent in charge of picking out specific programs from a Collection and inserting them into a policy map at a defined, predictable point during the loading process. All but one remaining (`cilium_calls_`) bpf_elf_map has been converted to BTF-style map definitions. The last step in sunsetting the iproute2 legacy is porting over cilium_calls, which is now well within reach.

ti-mo committed:
bpf,loader: explicitly manage policy programs for bpf_lxc and bpf_host

Remove the hack from loadDatapath that pulled out policy maps and programs
before loading a Collection. Explicitly commit them to the global policy maps
as part of loading the endpoint after getting all internal state in order first.

Also removes the POLICY_CALL_MAP macro to ensure stable map name in
CollectionSpec. POLICY_CALL_MAP would evaluate to test_cilium_policy_calls in tests,
while the agent would build the datapath with the macro set to cilium_policy_calls.
This is not compatible with the new way of loading ELFs, so the macro was removed.

The reason for the test_ map prefixes is not documented. With verifier
tests now unflagging maps for pinning, and the loader test suite taking
ownership of /sys/fs/bpf/{cilium,tc}, it doesn't look like there's a strong
reason for keeping the prefixes around. Removing all map name macros is a lot
of churn, so this will follow as its own change set at a later time.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: convert CUSTOM_CALLS_MAP to BTF map definition

This didn't need to be a bpf_elf_map for this long anyway, and it's anybody's
guess who's using this feature.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: move renameMaps into bpf.LoadCollection

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: unconditionally emit handle_policy_egress and egress policy map to ELF

A follow-up commit will assume these objects are always present in bpf_lxc.o.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: refactor loadDatapath into bpf.LoadAndAssign

This commit removes what's left of  loader.loadDatapath and moves it into the
bpf package, where it is pulled further apart and refactored into taking a
Map and Program receiver object. This automates away checking for Map and Program
objects in a Collection and adds a layer of runtime safety around it. Previously,
trying to access missing Maps or programs would panic.

LoadAndAssign remains the primary entry point for loader invocations from the
various subsystems (xdp, lxc, host, network, overlay) and retains its logging
behaviour and verifier error handling.

Loader behaviour remains largely unchanged.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: unconditionally emit handle_lxc_traffic to ELF

A follow-up commit will assume the program is always present in bpf_host.o.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: create and clean up well-known bpffs dirs in test suite

This is in preparation of a follow-up commit that removes the test_ prefix for
some maps.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: split pin commit mechanism into two stages

While introducing LoadAndAssign, an issue surfaced that resulted in commit()
failing on maps that were both marked for being replaced as well as being
assigned to a struct by L&A. Since commit() used to operate using string
references into Collection.Maps, and a prior Collection.Assign would delete()
entries from Collection.Maps, the commit operation would fail.

This patch splits the commit mechanism into two stages; before LoadCollection
returns, it pulls out real ebpf.Map objects from the Collection, allowing
a deferred commit() to call Map.Pin directly on the Map objects instead of
having to access the Collection.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: run docker target for checking testdata in CI

If the Makefile ends up building one of the .c files, it also invokes
llvm-objcopy for stripping the binary, which is not installed on the
GHA runners.

Always use the `docker` target for verifying if testdata was correctly
built, since different versions of the clang toolchain generate different
binaries. It's also important to run a clean build so Make doesn't simply
skip building the objects.

Signed-off-by: Timo Beckers <timo@isovalent.com>

bimmlerd commented on .gitignore:
```diff
@@ -72,6 +72,9 @@ test/cilium-[0-9a-f]*.yaml
 test/cilium-istioctl
 
+# Testdata, includes generated .o
+!**/testdata/*
```
Is there no way we can avoid committing these binary files?

ti-mo replied:
Good question! Stuffing binary files in Git is generally frowned upon, but:
- These files are generated by a single command using a stable, Dockerized toolchain
- CI always clean-builds the files to ensure there's no diff
- They're stripped as far down as possible, leaving just the ELF structure, BTF blob and the binary contents of the prog and map sections. The objects in this repo are max 4000 bytes large. For reference, this repo's vendor/ is currently a whopping 162 MiB of source code.

This buys us lightning-fast `go test` runs without any runtime toolchain or other external dependencies, a trade-off that's well worth it.

rgo3 commented on bpf/lib/maps.h:
```diff
@@ -134,14 +134,13 @@ struct bpf_elf_map __section_maps CALLS_MAP = {
-	.pinning	= LIBBPF_PIN_BY_NAME,
-	.max_elem	= 4,	/* ingress and egress, IPv4 and IPv6 */
-};
+struct {
```
Reading the respective commit message of the change, what's the risk for users when changing maps to a BTF definition?

ti-mo replied:
None! It's just a different representation in C, with the added benefit that the new definition will cause MapSpec.Key and .Value to be populated with btf.Type values corresponding to u32s, leading to `bpftool dump map` displaying pretty-prints. :)

ti-mo committed:
loader: separate host and lxc endpoint attachment

Avoid the nebulous term 'datapath' when describing individual bpf programs,
so rename `loader.reloadHostDatpath` to `loader.reloadHostEndpoint` and
`loader.reloadDatapath` to `loader.reloadEndpoint`.

For the host endpoint, the wireguard logic was lifted into reloadHost. Inserting
endpoint routes is omitted from the host endpoint attachment since endpoint routes
are not a thing for the host endpoint. Perform the ep.IsHost() check earlier in the
process in `loader.ReloadDatapath` to branch off into specific attachment code paths.

Aside from the above, no functional changes were made to the attachment process.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: break reloadHostEndpoint into multiple functions

This was getting a bit unwieldy, split into attachCiliumHost,
attachCiliumNet and attachNetworkDevices.

Signed-off-by: Timo Beckers <timo@isovalent.com>

rgo3 commented on pkg/datapath/loader/loader.go:
```diff
@@ -481,6 +469,23 @@ func (l *loader) reloadDatapath(ep datapath.Endpoint, spec *ebpf.CollectionSpec)
 		}
 		defer coll.Close()
 
+		// Insert policy programs before attaching entrypoints to tc hooks.
```
General question: During upgrade, does this update maps that are used by previously attached programs? And what are the risks of previously compiled tc attachments calling into newly compiled tail calls?

ti-mo replied:
It does, `obj.PolicyMap` points to the system-global `cilium_call_policy`, which the `.Update()` call puts an entry into. This doesn't only happen on upgrade, but on every endpoint regeneration (due to policy or config changes at runtime).

I think the code comment already provides a high-level description, but the risks of old Collections calling into new Collections heavily depend on the bpf C code in question, and I can't speak to that specifically. Programs in the policy map are called when e.g. the program evaluating cilium_host egress traffic towards a pod needs to do a policy check for that pod. This effectively crosses the boundary from bpf_host into a policy program living in bpf_lxc. When this happens, we need to make sure the target bpf_lxc's internal state (mostly its cilium_calls) is fully populated with the new programs so we don't hit any missed tail calls. As long as that's the case, we're good. :)

rgo3 reviewed (approved):
Overall LGTM, just two general questions about the code.  `LoadAndAssign` improves the readability by a lot :+1:
